### PR TITLE
feat: enable validation for contract calls

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/686.

## Changes

* Add a column to the `dkg_shares` for the number of signature shares that are necessary for a signature.
* Clarify some of the docstrings.
* Enable validation for stacks transactions by finishing up the function that does validation.
* Add a new validation check where each signer checks whether they are in the signer set.

## Testing Information

I added an integration test for the new validation check. I need to update one more integration test though.

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

File: protobufs/stacks/signer/v1/requests.proto [hunk 1]
```diff
@@ -16,7 +16,8 @@ message BitcoinTransactionSignRequest {
 
 // Represents a request to sign a Stacks transaction.
 message StacksTransactionSignRequest {
-  // The aggregate public key that will sign the transaction.
+  // This is the bitcoin aggregate key that was output from DKG. It is used
+  // to identifyy the signing set for the transaction.
```
Comment (AshtonStephens) [inline protobufs/stacks/signer/v1/requests.proto:20]:
```suggestion
  // to identify the signing set for the transaction.
```

File: signer/src/proto/generated/stacks.signer.v1.rs [hunk 2]
```diff
@@ -69,7 +69,8 @@ pub struct BitcoinTransactionSignRequest {
 #[allow(clippy::derive_partial_eq_without_eq)]
 #[derive(Clone, PartialEq, ::prost::Message)]
 pub struct StacksTransactionSignRequest {
-    /// The aggregate public key that will sign the transaction.
+    /// This is the bitcoin aggregate key that was output from DKG. It is used
+    /// to identifyy the signing set for the transaction.
```
Comment (AshtonStephens) [inline signer/src/proto/generated/stacks.signer.v1.rs:73]:
```suggestion
    /// to identify the signing set for the transaction.
```

File: signer/src/storage/model.rs [hunk 3]
```diff
@@ -339,6 +339,10 @@ pub struct EncryptedDkgShares {
     pub public_shares: Bytes,
     /// The set of public keys that were a party to the DKG.
     pub signer_set_public_keys: Vec<PublicKey>,
+    /// The threshold number of signature shares required to generate a
+    /// Schnorr signature.
```
Comment (AshtonStephens) [inline signer/src/storage/model.rs:343]: Can you add a little more detail for simpletons like me? Something like:

```
    /// The threshold number of signature shares required to generate a
    /// Schnorr signature, where a signature share is defined as  ...
```

File: signer/src/stacks/contracts.rs [hunk 4]
```diff
@@ -73,8 +73,8 @@ pub struct ReqContext {
     /// [`AsContractCall::validate`] function, but is here for logging and
     /// tracking purposes.
     pub origin: PublicKey,
-    /// The combined public key of all the signers in the current
-    /// signing set.
+    /// This is the bitcoin aggregate key that was output from DKG. It is
```
Comment (AshtonStephens) [inline signer/src/stacks/contracts.rs:76]: I think we should still specify that this is a public key in the description despite the fact it's clear from the type of the field as well.

```suggestion
    /// This is the bitcoin aggregate public key that was output from DKG. It is
```

File: signer/src/transaction_signer.rs [hunk 5]
```diff
@@ -463,7 +465,7 @@ where
         let multi_sig = MultisigTx::new_tx(&request.contract_call, &wallet, request.tx_fee);
         let txid = multi_sig.tx().txid();
 
-        // TODO: Make this more robust. The signer that recieves this won't
+        // TODO: Make this more robust. The signer that receives this won't
```
Comment (AshtonStephens) [inline signer/src/transaction_signer.rs:468]: If there's a ticket for this can you link it like `TODO(#12345):`, and if there isn't a ticket can you very quickly make a `tech debt` ticket? It should be fast, you can just put minimal details into it but we don't want to lose track.

File: signer/tests/integration/transaction_signer.rs [hunk 6]
```diff
@@ -359,3 +363,91 @@ async fn handle_pending_deposit_request_address_script_pub_key() {
...
+    // Let's get the blockchain data into the database.
+    let chain_tip: BitcoinBlockHash = setup.sweep_block_hash.into();
+    backfill_bitcoin_blocks(&db, rpc, &chain_tip).await;
+
+    // This is all normal happy path things that need to happen in order to
+    // pass validation.
+    setup.store_deposit_tx(&db).await;
+    setup.store_sweep_tx(&db).await;
+    setup.store_dkg_shares(&db).await;
+    setup.store_deposit_request(&db).await;
```
Comment (AshtonStephens) [inline signer/tests/integration/transaction_signer.rs:398]: Naive question; is this going to be necessary for a class of tests, or is this unique to this one? The comment makes me think it'll be necessary for a class of tests, at which point I think this is a good contender to be in something like a test setup function.

File: signer/tests/integration/transaction_signer.rs [hunk 7]
```diff
@@ -359,3 +363,91 @@ async fn handle_pending_deposit_request_address_script_pub_key() {
 
     testing::storage::drop_db(db).await;
 }
+
+/// Test that [`TxSignerEventLoop::assert_valid_stacks_tx_sign_request`]
+/// errors when the signer is not in the signer set.
+#[cfg_attr(not(feature = "integration-tests"), ignore)]
+#[tokio::test]
+async fn signing_set_validation_check_for_stacks_transactions() {
```
Comment (AshtonStephens) [inline signer/tests/integration/transaction_signer.rs:371]: Love the comments on this tests steps :)

File: signer/tests/integration/transaction_signer.rs [hunk 8]
```diff
@@ -359,3 +363,91 @@ async fn handle_pending_deposit_request_address_script_pub_key() {
...
+    // Let's create a proper sign request.
+    let request = StacksTransactionSignRequest {
+        aggregate_key: setup.aggregated_signer.keypair.public_key().into(),
+        contract_call: signer::stacks::contracts::ContractCall::CompleteDepositV1(req),
+        // The nonce and tx_fee aren't really validated against anything at
+        // the moment.
+        nonce: 1,
+        tx_fee: 100_000,
+        // This can probably be removed, but it's not important to remove
+        // now, we should switch to protobuf messages and remove it then.
```
Comment (AshtonStephens) [inline signer/tests/integration/transaction_signer.rs:425]: This comment might be fine, but if you haven't already consider if this is worth making a tech-debt type issue for to track, and if so it would be good to link it here with a `TODO`

File: signer/src/wsts_state_machine.rs [hunk 9]
```diff
@@ -148,13 +148,19 @@ impl SignerStateMachine {
             wsts::util::encrypt(&self.0.network_private_key.to_bytes(), &encoded, rng)
                 .map_err(|_| error::Error::Encryption)?;
 
+        let signature_share_threshold: u16 = self
+            .threshold
+            .try_into()
+            .map_err(|_| Error::TypeConversion)?;
```
Comment (AshtonStephens) [inline signer/src/wsts_state_machine.rs:154]: Not a nit, just a question: 

I see you mostly do the type casts explicitly at the moment that something could fail instead of having a `From` definition for the error that the function returns. Why?

File: protobufs/stacks/signer/v1/requests.proto (hunk 1 above)
Comment (djordon) [reply protobufs/stacks/signer/v1/requests.proto:20]: Oh man, for some reason I thought I got all of these.

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:468]: We don't. And I should remove or reword this TODO.

I'm leaning toward removing the digest field since it will be unnecessary if we just switch to signing the entire message bytes. This work is basically https://github.com/stacks-network/sbtc/issues/517, and is super easy to do. In fact, I will do it now.

File: signer/tests/integration/transaction_signer.rs (hunk 8 above)
Comment (djordon) [reply signer/tests/integration/transaction_signer.rs:425]: I'm going to fix #517 and remove this.

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:468]: Actually, I won't do this now, and just update the todo.

File: signer/tests/integration/transaction_signer.rs (hunk 8 above)
Comment (djordon) [reply signer/tests/integration/transaction_signer.rs]: Actually, I'm just going to add a TODO https://github.com/stacks-network/sbtc/pull/694/commits/c3ceea7f8a43bcdd8ef975170e6c5d9aaf965461.

File: signer/src/transaction_signer.rs (hunk 5 above)
Comment (djordon) [reply signer/src/transaction_signer.rs]: Added a TODO and reworded this in https://github.com/stacks-network/sbtc/pull/694/commits/c3ceea7f8a43bcdd8ef975170e6c5d9aaf965461.

File: signer/src/stacks/contracts.rs (hunk 4 above)
Comment (djordon) [reply signer/src/stacks/contracts.rs]: https://github.com/stacks-network/sbtc/pull/694/commits/c3ceea7f8a43bcdd8ef975170e6c5d9aaf965461

File: signer/src/storage/model.rs (hunk 3 above)
Comment (djordon) [reply signer/src/storage/model.rs:343]: Done in https://github.com/stacks-network/sbtc/pull/694/commits/c3ceea7f8a43bcdd8ef975170e6c5d9aaf965461.

File: signer/src/proto/generated/stacks.signer.v1.rs (hunk 2 above)
Comment (djordon) [reply signer/src/proto/generated/stacks.signer.v1.rs]: https://github.com/stacks-network/sbtc/pull/694/commits/c3ceea7f8a43bcdd8ef975170e6c5d9aaf965461

File: signer/src/wsts_state_machine.rs (hunk 9 above)
Comment (djordon) [reply signer/src/wsts_state_machine.rs:154]: I tend to do this for a few reasons:
1. The error returned from the `try_into` is not particularly informative, so having the dedicated variant is not super helpful.
2. I think implementing `From` for an error variant is not the right pattern for us to have. Instead the goal is to construct error variants with enough context data for us to debug when we encounter the error. And `From` impls, particularly `From` impls for foreign types, encourages the wrong behavior.
3. This particular error is not actually "real". The threshold is a `u32` type, but it is created from a `u16`, so the conversion back should always happen. So I don't think it's necessary to add context and all of that, I'd rather get rid of the error.

Because of (3) I could do an `as u16` cast, but I'm hesitant because it's kind of bad practice (as in we should be cautious of using them). Perhaps it's warranted in this case with a comment explaining why it is fine.

File: signer/tests/integration/transaction_signer.rs (hunk 6 above)
Comment (djordon) [reply signer/tests/integration/transaction_signer.rs]: Done in https://github.com/stacks-network/sbtc/pull/694/commits/4870b46563051b1fcf6805c1780cfd37f0bfa7a2.
